src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -607,6 +589,47 @@ private void AssertNotDisposed()
 #pragma warning restore CA1513 // Use ObjectDisposedException throw helper
     }
 
+    private ValueTask InvokeCircuitHandlersAsync<T1, T2>(string methodName, Func<CircuitHandler, Func<T1, T2, Task>> getMethod, T1 arg1, T2 arg2)
+        => InvokeCircuitHandlersCoreAsync(methodName, static (handler, args) =>
+        {
+            var invokeHandler = args.getMethod(handler);
+            return invokeHandler(args.arg1, args.arg2);
+        }, (getMethod, arg1, arg2));
+
+    private ValueTask InvokeCircuitHandlersAsync<T1, T2, T3>(string methodName, Func<CircuitHandler, Func<T1, T2, T3, Task>> getMethod, T1 arg1, T2 arg2, T3 arg3)
+        => InvokeCircuitHandlersCoreAsync(methodName, static (handler, args) =>
+        {
+            var invokeHandler = args.getMethod(handler);
+            return invokeHandler(args.arg1, args.arg2, args.arg3);
+        }, (getMethod, arg1, arg2, arg3));
+
+    // Note: we log exceptions and re-throw while running handlers, because there may be multiple
+    // exceptions.
+    private async ValueTask InvokeCircuitHandlersCoreAsync<TArgs>(string methodName, Func<CircuitHandler, TArgs, Task> invokeHandler, TArgs args)
```

Comment:
The goal behind all this was to avoid copying and pasting the same ~20 lines of code each time we wanted to invoke a handler method. These helpers are a way of abstracting away the duplicated logic and avoiding creating a closure each time a handler gets invoked. Maybe this is an over-optimization and we're actually okay with slightly simpler-looking code at the cost of a few additional allocations.

I'm guessing the potential clumsiness of abstracting some of this out was the reason we took the copy/paste approach before, but the problem is only going to get worse the more `CircuitHandler` methods we add.